module/util/util.go

```diff
@@ -49,3 +50,60 @@ func AllDone(components ...module.ReadyDoneAware) <-chan struct{} {
 
 	return done
 }
+
+// WaitReady waits for either a signal/close on the ready channel or for the context to be cancelled
+// Returns nil if the channel was signalled/closed before returning, otherwise, it returns the context
+// error.
+//
+// This handles the corner case where the context is cancelled at the same time that components
+// were marked ready, and the Done case was selected.
+// This is intended for situations where ignoring a signal can cause safety issues.
+func WaitReady(ctx context.Context, ready <-chan struct{}) error {
```

Comment:
nit: since this is just an arbitrary channel, might be better to just define it as:

```suggestion
func WaitClosed(ctx context.Context, ch <-chan struct{}) error {
```

Reply:
yea, that's a better name